jlm/llvm/ir/operators/Load.hpp

```diff
@@ -575,6 +577,134 @@ class LoadNonVolatileNode final : public LoadNode
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadAlloca(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief Forwards the value from a store operation.
+ *
+ * s2 = store_op a v1 s1
+ * v2 s3 = load_op a s2
+ * ... = any_op v2
+ * =>
+ * s2 = store_op a v1 s1
+ * ... = any_op v1
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadStore(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief If the producer of a load's address is an alloca operation, then we can remove all
+ * state edges originating from other alloca operations coming through store operations.
+ *
+ * a1 sa1 = alloca_op ...
+ * a2 sa2 = alloca_op ...
+ * ss1 = store_op a1 ... sa1
+ * ss2 = store_op a2 ... sa2
+ * ... = load_op a1 ss1 ss2
+ * =>
+ * ...
+ * ... = load_op a1 ss1
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadStoreState(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief Remove duplicated state operands
+ *
+ * v so1 so2 so3 = load_op a si1 si1 si1
+ * =>
+ * v so1 = load_op a si1
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadDuplicateState(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief Avoid sequentialization of load operations.
+ *
+ * _ so1 = load_op _ si1
+ * _ so2 = load_op _ so1
+ * _ so3 = load_op _ so2
+ * =>
+ * _ so1 = load_op _ si1
+ * _ so2 = load_op _ si1
+ * _ so3 = load_op _ si1
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadLoadState(
```

Comment:
Does this normalization create a memory state merge node? A later store that depends on `so3` needs to be serialized after all three loads, so it can no longer just depend on `so3` in the new graph.

Reply:
Yes, it creates one: https://github.com/phate/jlm/blob/5c82eb176c6892c9a0528cb0593decd1e8e8be5d/jlm/llvm/ir/operators/Load.cpp#L554C1-L564C4